snapcraft_legacy/storeapi/v2/validation_sets.py

```diff
@@ -14,161 +14,119 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-from typing import Any, Dict, List, Optional
+from typing import Any, Annotated, Dict, Literal, TYPE_CHECKING
+import pydantic
 
-import jsonschema
+from craft_application import models
 
-from ._api_schema import BUILD_ASSERTION_JSONSCHEMA
 
+if TYPE_CHECKING:
+    SnapName = str
+    SnapId = str
+else:
+    SnapName = pydantic.constr(max_length=40)
+    SnapId = pydantic.constr(max_length=40)
 
-class Snap:
+
+def cast_dict_values_to_strings(data: dict[str, Any]) -> dict[str, Any]:
+    """Cast all scalars in a dictionary to strings."""
+    return {k: _to_string(v) for k, v in data.items()}
+
+
+def _to_string(data: list | str | int | None) -> dict | list | str | None:
+    """Recurse through nested lists and cast all scalars to strings."""
+    if isinstance(data, dict):
+        return {k: _to_string(v) for k, v in data.items()}
+    elif isinstance(data, list):
+        return [_to_string(i) for i in data]
+    elif data:
+        return str(data)
+    return data
```

Comment:
`_to_string(0)` seems like it would output `0` rather than `"0"`

Reply:
Good point.  I changed the code to look for common scalars (int, float, bool) and updated the docstring to reflect that.